gector/gec_predictor.py

```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
+                    "logits_d_tags": None,
+                    "class_probabilities_labels": None,
+                    "class_probabilities_d_tags": None,
+                    "max_error_probability": None,
+                    "words": instance.fields["tokens"].tokens[1:],
+                    "labels": None,
+                    "d_tags": None,
+                    "corrected_words": instance.fields["tokens"].tokens[1:],
+                }
+            else:
+                pred_ids.append(id)
+
+        # Applying correction model multiple times
+        for _ in range(self._iterations):
+
+            # If no sentences need to be passed into model
+            if len(pred_ids) == 0:
+                break
+
+            # Create batch of instances to be passed into model
+            orig_batch = [final_batch[pred_id] for pred_id in pred_ids]
+
+            # Pass into model
+            outputs = self._model.forward_on_instances(orig_batch)
+
+            new_pred_ids = []
+
+            # Output_ID and Pred_ID in pred_ids
+            for op_ind, pred_id in enumerate(pred_ids):
+
+                # Update final outputs
+                final_outputs[pred_id] = outputs[op_ind]
+                orig = final_batch[pred_id]
+
+                # Create tokens from corrected words for next iter
+                tokens = [
+                    Token(word)
+                    for word in ["$START"] + outputs[op_ind]["corrected_words"]
+                ]
+
+                # Tokens to instance
+                pred = self._dataset_reader.text_to_instance(tokens)
+                prev_preds = prev_preds_dict[pred_id]
+
+                # If model output is different from previous iter outputs
+                # Update input batch, append to dict and add to `pred_ids`
+                if (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
+                    and pred.fields["tokens"].tokens not in prev_preds
+                ):
+                    # Set sentence instance to predicted sent in batch
+                    final_batch[pred_id] = pred
+                    # Append orig_id in new_pred_ids
+                    new_pred_ids.append(pred_id)
+                    # Append new prediction to prev_preds_dict
+                    prev_preds_dict[pred_id].append(
+                        pred.fields["tokens"].tokens
+                    )
+                # If model output is same as that in prev iter, update final batch
+                # but stop passing it into the model for future iters
+                # This means that no corrections have been made in this iteration
+                elif (
+                    orig.fields["tokens"].tokens != pred.fields["tokens"].tokens
+                    and pred.fields["tokens"].tokens in prev_preds
+                ):
+                    # update final batch, but stop iterations
+                    final_batch[pred_id] = pred
```

Comment:
So we set final_batch[i] in cases where i was pointing to an instance with 4 or fewer words, but it doesn't seem like those cases are skipped here. Wouldn't that cause final_batch[i] to be overwritten for these short cases? Is this an error in gec_model that we're replicating for consistency?

Reply:
This is not the case for short sentences. Such sentences are not passed into the model at all. GEC_Model has a min_length of 3.

This is if no corrections have been made to the sentence in this iteration. So the list of `pred_ids` is not updated for these sentences.